hugolib/hugo_sites.go

```diff
@@ -822,7 +822,8 @@ func (h *HugoSites) loadData(fis []hugofs.FileMetaInfo) (err error) {
 			return err
 		}
 		for _, r := range files {
-			if err := h.handleDataFile(r); err != nil {
+			// load the data file only if its content is not empty.
+			if err := h.handleDataFile(r); r.FileInfo().Size() != 0 && err != nil {
```

Comment:
I'll update the `if` statement into a `nested if` to first check for the size in the loop and continue if 0.

For the test I suppose I can write a function (say) `TestDataDirEmptyFiles` within [datafiles_test.go](https://github.com/gohugoio/hugo/blob/master/hugolib/datafiles_test.go)  as below : 

```go
func TestDataDirEmptyFiles(t *testing.T) {
	t.Parallel()
	equivDataDirs := make([]dataDir, 3)
	equivDataDirs[0].addSource("data/test/a.json", ``)
	equivDataDirs[1].addSource("data/test/a.yaml", "")
	equivDataDirs[2].addSource("data/test/a.toml", "")
	expected := map[string]interface{}{}
	doTestEquivalentDataDirs(t, equivDataDirs, expected)
}
```
Is that the correct approach ?

Reply:
I have updated the fix for the issue [here](https://github.com/gohugoio/hugo/pull/8603/files).

Instead of not loading empty data files. In this fix, we are loading all data files including empty ones. 
But we are setting the `content` for empty data files to `nil` to avoid unmarshal error which causes a crash in #8601

With this approach we are also able to use `len` for empty data files which was not possible with the previous approach.

Expected Behaviour :
If you add one/more 0 byte json files to your data directory which may or maynot be referenced : `no errors`

I have manually tested it multiple times. But I'm not sure how to write a test for it.

Reply:
I have updated the fix [here](https://github.com/gohugoio/hugo/pull/8603/commits/0fd028893bfa11a764b9b34cbc0500bd9281ab3a).

```diff
- if data == nil {
+ if len(data) == 0 {
```
`data == nil` yields `false` for empty as well as non empty data files. So the if statement which is supposed to return an empty map for empty data files is never reached. Hence the crash with `unmarshal failed` error.

however, `len(data) == 0` yields `true` only for empty data files ergo the if block takes care of empty the data files as intended.